local-cluster/src/local_cluster.rs

```diff
@@ -665,6 +668,53 @@ impl LocalCluster {
         info!("{} done waiting for roots", test_name);
     }
 
+    /// Attempt to send and confirm tx "attempts" times
+    /// Wait for signature confirmation before returning
+    /// Return the transaction signature
+    pub fn send_transaction_with_retries<T: Signers + ?Sized>(
+        client: &QuicTpuClient,
+        keypairs: &T,
+        transaction: &mut Transaction,
+        attempts: usize,
+        pending_confirmations: usize,
+    ) -> std::result::Result<Signature, TransportError> {
+        for attempt in 0..attempts {
+            let now = Instant::now();
+            let mut num_confirmed = 0;
+            let mut wait_time = MAX_PROCESSING_AGE;
+
+            while now.elapsed().as_secs() < wait_time as u64 {
+                if num_confirmed == 0 {
+                    client.send_transaction_to_upcoming_leaders(transaction)?;
+                }
+
+                if let Ok(confirmed_blocks) = client.rpc_client().poll_for_signature_confirmation(
+                    &transaction.signatures[0],
+                    pending_confirmations,
+                ) {
+                    num_confirmed = confirmed_blocks;
+                    if confirmed_blocks >= pending_confirmations {
+                        return Ok(transaction.signatures[0]);
+                    }
+                    // Since network has seen the transaction, wait longer to receive
+                    // all pending confirmations. Resending the transaction could result into
+                    // extra transaction fees
+                    wait_time = wait_time.max(
+                        MAX_PROCESSING_AGE * pending_confirmations.saturating_sub(num_confirmed),
+                    );
+                }
+            }
+            info!("{attempt} tries failed transfer");
+            let blockhash = client.rpc_client().get_latest_blockhash()?;
+            transaction.sign(keypairs, blockhash);
+        }
+        Err(std::io::Error::new(
+            std::io::ErrorKind::Other,
+            "failed to confirm transaction".to_string(),
+        )
+        .into())
+    }
```

Comment:
this part is similar to thin_client and I think it is closer to the original behavior which before we switched to QuicTpuClient.
https://github.com/anza-xyz/agave/blob/a5229f989ca47520f6ed62c2d3503d5e84da093e/thin-client/src/thin_client.rs#L214-L261

I'm a little bit confused why it compares `MAX_PROCESSING_AGE` with `now.elapsed().as_secs()` but I think we can update this one in different PR!

Reply:
Valid concern. The idea of the original code is that if we waited longer than `MAX_PROCESSING_AGE` it means that the blockhash has expired so we must update blockhash and resign which might a desired behavior for production applications.
Yet  I'm concerned about using this code for integration tests since in the worst case scenario we can wait for minutes for transaction to be delivered, not sure if it worth it or better to accept failure if we cannot send within several leader slots (opposite to hundreds).